src/kqueue.rs

```diff
@@ -0,0 +1,380 @@
+            .into_iter()
+            .filter_map(filter_dir)
+        {
+            self.add_single_watch(entry.path().to_path_buf(), is_recursive)?;
+        }
+        self.kqueue.watch()?;
+
+        Ok(())
+    }
+
+    fn add_single_watch(&mut self, path: PathBuf, is_recursive: bool) -> Result<()> {
+        let event_filter = EventFilter::EVFILT_VNODE;
+        let filter_flags = FilterFlag::NOTE_DELETE
+            | FilterFlag::NOTE_WRITE
+            | FilterFlag::NOTE_EXTEND
+            | FilterFlag::NOTE_ATTRIB
+            | FilterFlag::NOTE_LINK
+            | FilterFlag::NOTE_RENAME
+            | FilterFlag::NOTE_REVOKE;
+
+        self.kqueue
+            .add_filename(&path, event_filter, filter_flags)?;
+        self.watches.insert(path, is_recursive);
+        self.kqueue.watch()?;
+        Ok(())
+    }
+
+    fn remove_watch(&mut self, path: PathBuf, remove_recursive: bool) -> Result<()> {
+        match self.watches.remove(&path) {
+            None => return Err(Error::watch_not_found()),
+            Some(is_recursive) => {
+                self.kqueue
+                    .remove_filename(&path, EventFilter::EVFILT_VNODE)
+                    .map_err(|e| Error::io(e))?;
+
+                if is_recursive || remove_recursive {
+                    for entry in WalkDir::new(path)
+                        .follow_links(true)
+                        .into_iter()
+                        .filter_map(filter_dir)
+                    {
+                        self.kqueue.remove_filename(
+                            entry.path().to_path_buf(),
+                            EventFilter::EVFILT_VNODE,
+                        )?;
+                    }
+                }
+                self.kqueue.watch()?;
+            }
+        }
+        Ok(())
+    }
+}
+
+/// return `DirEntry` when it is a directory
+fn filter_dir(e: walkdir::Result<walkdir::DirEntry>) -> Option<walkdir::DirEntry> {
+    if let Ok(e) = e {
+        if let Ok(metadata) = e.metadata() {
+            if metadata.is_dir() {
+                return Some(e);
+            }
+        }
+    }
+    None
+}
+
+impl KqueueWatcher {
+    fn from_event_fn(event_fn: Box<dyn EventFn>) -> Result<Self> {
+        let kqueue = kqueue::Watcher::new()?;
+        let event_loop = EventLoop::new(kqueue, event_fn)?;
+        let channel = event_loop.event_loop_tx.clone();
+        let waker = event_loop.event_loop_waker.clone();
+        event_loop.run();
+        Ok(KqueueWatcher { channel, waker })
+    }
+
+    fn watch_inner(&mut self, path: &Path, recursive_mode: RecursiveMode) -> Result<()> {
+        let pb = if path.is_absolute() {
+            path.to_owned()
+        } else {
+            let p = env::current_dir().map_err(Error::io)?;
+            p.join(path)
+        };
+        let (tx, rx) = unbounded();
+        let msg = EventLoopMsg::AddWatch(pb, recursive_mode, tx);
+
+        // we expect the event loop to live and reply => unwraps must not panic
```

Comment:
I think we could also just report this via the Result and not panic randomly inside application code. But that's not a hard opinion, just an idea.

Reply:
This is also copied the INotifyWatcher

Reply:
After looking over this peace of code, i agree with you. To which `ErrorKind` should i map this, or should this maybe be resolved in an another  PR where this behavior is also removed from `INotifyWatcher`.

Reply:
error::Generic is fine